presto-hive/src/main/java/com/facebook/presto/hive/HiveFileRenamer.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.facebook.presto.spi.ConnectorMetadataUpdateHandle;
+import com.facebook.presto.spi.SchemaTableName;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import org.weakref.jmx.Managed;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class HiveFileRenamer
+{
+    private final LoadingCache<HiveMetadataUpdateKey, AtomicLong> partitionFileCounterCache;
+    private final LoadingCache<HiveMetadataUpdateHandle, String> hiveMetadataResultCache;
```

Comment:
We should use a concurrent hash map or something like to to remember everything. We purged finished queries by adding a callback to `QueryManager.addStateChangeListener`. Note that the callback itself could unfortunately introduce another interface in connector metadata. Let me know if you might come up with better solution than that.

Reply:
How about we inject `QueryManager` directly into `HiveFileRenamer` because this is to help purges entries in `HiveFileRenamer` specifically. WDYT ?